File: Source/Plugins/BulkWorkflow/com.tle.web.bulk.workflow/src/com/tle/web/bulk/workflow/operations/WorkflowMoveOperation.java
package com.tle.web.bulk.workflow.operations;

import javax.inject.Inject;

import com.dytech.edge.exceptions.WorkflowException;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.assistedinject.AssistedInject;
import com.tle.annotation.Nullable;
import com.tle.beans.item.HistoryEvent;
import com.tle.beans.item.HistoryEvent.Type;
import com.tle.beans.item.ModerationStatus;
import com.tle.common.i18n.CurrentLocale;
import com.tle.common.workflow.WorkflowNodeStatus;
import com.tle.common.workflow.node.ParallelNode;
import com.tle.common.workflow.node.WorkflowItem;
import com.tle.common.workflow.node.WorkflowNode;
import com.tle.core.item.standard.ItemOperationFactory;
import com.tle.core.item.standard.operations.workflow.TaskOperation;
import com.tle.core.item.standard.workflow.nodes.ScriptStatus;
import com.tle.core.item.standard.workflow.nodes.TaskStatus;
import com.tle.core.security.TLEAclManager;
import com.tle.exceptions.AccessDeniedException;

public class WorkflowMoveOperation extends TaskOperation
{
	private final String msg;
	private final String toStep;

	@Inject
	private ItemOperationFactory workflowFactory;
	@Inject
	private TLEAclManager aclService;

	@AssistedInject
	public WorkflowMoveOperation(@Assisted("msg") String msg, @Assisted("toStep") @Nullable String toStep)
	{
		this.msg = msg;
		this.toStep = toStep;
	}

	@Override
	public boolean execute()
	{
		if( !aclService.checkPrivilege("MANAGE_WORKFLOW", getWorkflow()) )
		{
			throw new AccessDeniedException(CurrentLocale.get("com.tle.core.services.item.error.nopriv",
				"MANAGE_WORKFLOW", getItemId()));
		}
		
		ModerationStatus status = getModerationStatus();
		status.setLastAction(params.getDateNow());

		int count = 0;
		boolean isMovingBack = false;
		for( WorkflowNodeStatus beanstatus : status.getStatuses() )
		{
			if( beanstatus.getNode().getType() == WorkflowNode.ITEM_TYPE
				&& beanstatus.getStatus() == WorkflowNodeStatus.INCOMPLETE )
			{
				TaskStatus task = (TaskStatus) getNodeStatus(beanstatus.getNode().getUuid());
				params.setCause(task.getBean());
				HistoryEvent move = createHistory(Type.taskMove);
				move.setComment(msg);
				setToStepFromTask(move, toStep);
				setStepFromTask(move, task.getId());
				WorkflowNode targetNode = task.getToStepByTaskId(toStep);

				if( targetNode == null )
				{
					throw new WorkflowException(
						CurrentLocale.get("com.tle.web.bulk.workflow.bulkop.movetask.error.nostep"));
				}

				if( isMovingToNewNode(targetNode) )
				{
					WorkflowNode parent = targetNode.getParent();
					if( parent != null && parent instanceof ParallelNode )
					{
						enter(parent, true);
					}
					else
					{
						enter(targetNode, true);
					}

					beanstatus.setStatus(WorkflowNodeStatus.COMPLETE);
					itemService.operation(getItemId(), workflowFactory.save());
				}
				else
				{
					if( count == 0 )
					{
						WorkflowNode parent = targetNode.getParent();
						if( parent != null && parent instanceof ParallelNode )
						{
							if( beanstatus.getNode().getParent() == parent )
							{
								continue;
							}

							enter(parent);
							beanstatus.setStatus(WorkflowNodeStatus.ARCHIVED);
							isMovingBack = true;
						}
						else
						{
							enter(targetNode);
							if( !task.isMovingTaskForward(beanstatus.getNode().getUuid(), toStep) )
							{
								beanstatus.setStatus(WorkflowNodeStatus.ARCHIVED);
							}
						}
					}
					else
					{
						if( isMovingBack || !task.isMovingTaskForward(beanstatus.getNode().getUuid(), toStep) )
						{
							beanstatus.setStatus(WorkflowNodeStatus.ARCHIVED);
						}
						else
						{
							beanstatus.setStatus(WorkflowNodeStatus.COMPLETE);
							itemService.operation(getItemId(), workflowFactory.save());
						}
					}
				}
				exitTask((WorkflowItem) beanstatus.getNode());
				count++;
			}
			else
				if( beanstatus.getNode().getType() == WorkflowNode.SCRIPT_TYPE
					&& beanstatus.getStatus() == WorkflowNodeStatus.INCOMPLETE )
			{
				ScriptStatus task = (ScriptStatus) getNodeStatus(beanstatus.getNode().getUuid());
				params.setCause(task.getBean());
				HistoryEvent move = createHistory(Type.taskMove);
				move.setComment(msg);
				setToStepFromTask(move, toStep);
				setStepFromTask(move, task.getId());
				WorkflowNode targetNode = task.getToStepByTaskId(toStep);
				reenter(targetNode);
			}
		}
		updateModeration();
		return true;
	}
}

File: Source/Plugins/BulkWorkflow/com.tle.web.bulk.workflow/scalasrc/com/tle/web/bulk/workflow/operations/WorkflowMoveOperation.scala
/*
 * Copyright 2017 Apereo
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tle.web.bulk.workflow.operations

import java.util.Date
import javax.inject.Inject

import com.google.inject.assistedinject.{Assisted, AssistedInject}
import com.tle.annotation.Nullable
import com.tle.beans.item.HistoryEvent
import com.tle.common.i18n.CurrentLocale
import com.tle.common.workflow.{WorkflowItemStatus, WorkflowNodeStatus}
import com.tle.common.workflow.node.WorkflowNode
import com.tle.core.item.NodeStatus
import com.tle.core.item.standard.ItemOperationFactory
import com.tle.core.item.standard.operations.workflow.TaskOperation
import com.tle.core.security.TLEAclManager
import com.tle.core.security.impl.SecureInModeration
import com.tle.exceptions.AccessDeniedException

import scala.collection.JavaConverters._

@SecureInModeration
class WorkflowMoveOperation @AssistedInject()(@Assisted("msg") val msg: String, @Assisted("toStep") val toStep: String) extends TaskOperation {

  @Inject var aclService: TLEAclManager = _


  override def execute: Boolean = {
    if (!aclService.checkPrivilege("MANAGE_WORKFLOW", getWorkflow)) {
      throw new AccessDeniedException(CurrentLocale.get("com.tle.core.services.item.error.nopriv", "MANAGE_WORKFLOW", getItemId))
    }

    clearAllStatuses()

    val nodeSeq = getWorkflow.getNodes.asScala.toSeq

    def requiresSiblingCompletion(wn: WorkflowNode) = wn.getType != WorkflowNode.PARALLEL_TYPE

    val parentMap = nodeSeq.groupBy(n => Option(n.getParent))

    def newStatus(completed: Boolean)(n: WorkflowNode) = {
      val ns = n.getType match {
        case WorkflowNode.ITEM_TYPE => val wis = new WorkflowItemStatus(n, null)
          wis.setStarted(params.getDateNow)
          wis
        case _ => new WorkflowNodeStatus(n)
      }
      ns.setStatus(if (completed) WorkflowNodeStatus.COMPLETE else WorkflowNodeStatus.INCOMPLETE)
      ns
    }

    nodeSeq.find(_.getUuid == toStep).foreach { wn =>

      def completePreviousSiblings(child: WorkflowNode): Seq[WorkflowNodeStatus] = (for {
        parent <- Option(child.getParent).filter(requiresSiblingCompletion)
        children <- parentMap.get(Some(parent))
      } yield {
        children.filter(_.getChildIndex < child.getChildIndex).map(newStatus(true))
      }).getOrElse(Seq.empty)

      def addParentStatuses(child: WorkflowNode): Seq[WorkflowNodeStatus] = {
        (completePreviousSiblings(child) :+ newStatus(false)(child)) ++
          Option(child.getParent).toSeq.flatMap(addParentStatuses)
      }

      val newStatuses = addParentStatuses(wn)
      initStatusMap(newStatuses.asJava)
      newStatuses.foreach { wns =>
        if (wns.getStatus == WorkflowNodeStatus.INCOMPLETE && !wns.getNode.isLeafNode) {
          update(wns.getNode)
        }
      }
      enter(wn)
      val h = createHistory(HistoryEvent.Type.taskMove)
      setToStepFromTask(h, toStep)
      h.setComment(msg)
    }
    updateModeration()
    true
  }
}
